File: satellite.py
import math

class satellite:
    """
    卫星类，默认以 STARLINK-31328 - 轨道参数
    a:半长轴
    b:偏心率
    i:轨道倾角
    w:近地点幅角
    Omega0:初始时刻升交点赤经
    M0:初始时刻平近点角
    bOmega:升交点赤经变化率
    bM:平近点角变化率
    G0:起始时刻格林威治恒星时角
    we:地球自转速度

    theta:卫星相位
    t:卫星现在运行时刻

    Omega:升交点赤经
    M:平近点角

    """
    def __init__(
        self,
        a=481.5,
        e=0.0001325,
        i=53.1557,
        w=81.8308,
        Omega0=182.3353,
        M0=325.1008,
        bOmega=0.001,
        bM=1,
        G0=150,
        omegae=np.degrees(7.23e-5)
        ):
        self.M0=M0
        self.Omega0=Omega0
        self.bM=bM
        self.bOmega=bOmega
        self.i=i
        self.G0=G0
        self.omegae=omegae
        # 初始时刻赤经和赤纬
        self.delta=math.asin(math.sin(math.radians(self.M0))*math.sin(math.radians(i)))
        self.alpha=Omega0+math.atan(math.tan(math.radians(self.M0))*math.cos(math.radians(i)))
        # 初始时刻星下点地心经纬度
        self.lamda=self.alpha-(G0+omegae*0)
        self.phi=self.delta



    def update(self,theta,t):
        """
        更新卫星在特定相位和特定时间下的星下点
        """
        self.M=self.M0+theta+self.bM*t
        self.Omega=self.Omega0+self.bOmega*t

        # self.delta=math.asin((math.sin(self.M))*math.sin(self.i))
        # self.alpha=self.Omega+math.atan(math.tan(self.M)*math.cos(self.i))
        self.delta=math.asin((math.sin(math.radians(self.M)))*math.sin(math.radians(self.i)))
        self.alpha=self.Omega+math.atan(math.tan(math.radians(self.M))*math.cos(math.radians(self.i)))

        self.lamda=self.alpha-(self.G0+self.omegae*t)
        self.phi=self.delta

File: Satellite.m
classdef Satellite
    properties
        days,h;
        a,e,i,Omega,omega,M;
        G_0,b_M,b_Omega;
        lambda_T,phi_T,rho_bar;
        tau;
    end
    methods
        % Inital methods
        function obj=Satellite(theta0)
            obj=set_orbit_parameters(obj,theta0);
            obj=set_time_parameters(obj);
            obj=set_auxiliary_parameters(obj);
            obj=cal_auxiliary_parameters(obj);
            obj=set_target(obj);

            obj=cal_tau(obj);
        end
        function obj=set_orbit_parameters(obj,M,a,e,i,Omega,omega)
            % Allow setup orbital elements without defaults.
            %
            % Parameters：
            % M: (float)Initial mean anomaly(deg).
            % a: (float,optional) Semi-major axis (m). Defaults to 6939.76e3.
            % e: (float,optional) eccentricity. Defaults to 0.
            % i: (float,optional) Orbital inclination(deg). Defaults to 98.
            % Omega: (float,optional) Right ascension of ascending node(deg). Defaults to 284.507.
            % omega: (float,optional) Argument of perigee (deg). Defaults to 0.

            % Verify Parameters
            arguments
                obj
                M (1,1) double
                a (1,1) double = 6939.76e3
                e(1,1) double=0
                i(1,1) double=98
                Omega(1,1) double=284.507
                omega(1,1) double=0
            end
            obj.M=M;
            obj.a=a;
            obj.e=e;
            obj.i=i;
            obj.Omega=Omega;
            obj.omega=omega;
        end
        function obj=set_time_parameters(obj,days,h)
            % Allow setup time parameters without defaults.
            %
            % Parameters：
            % days: (float,optional)total days of simulation. Defaults to 1.
            % h: (float,optional)Discrete interval (s). Defaults to 5.
            %
            % Verify Parameters
            arguments
                obj
                days (1,1) double = 1
                h (1,1) double = 5
            end
            obj.days=days;
            obj.h=h;
        end
        function obj=set_auxiliary_parameters(obj,G_0)
            % Allow setup auxiliary parameters without defaults.
            % 
            % Parameters：
            % G_0: (float,optional)total days of simulation. Defaults to 1. 
            % 
            % Verify Parameters
            arguments
                obj
                G_0 (1,1) double = 0
            end
            obj.G_0=G_0;
        end
        function obj=cal_auxiliary_parameters(obj)
            % Calculate auxiliary parameters b_Omega and b_M.
            miu=3.9860044e14;
            J2=1.08263e-3;
            R_e=6371.393e3;
            n_M=sqrt(miu/obj.a^3);
            CJ2=1.5*n_M*J2*(R_e/obj.a)^2;
            i_temp=deg2rad(obj.i);
            obj.b_Omega=-CJ2*cos(i_temp);
            obj.b_M=n_M-CJ2*(1.5*sin(i_temp)-1);
        end
        function obj=set_target(obj,lambda_T,phi_T,rho_bar)
            % Allow setup target latitude and longitudes without defaults.
            % Verify Parameters
            arguments
                obj
                lambda_T (1,1) double = 121.3
                phi_T (1,1) double = 31.1
                rho_bar (1,1) double = 10
            end
            obj.lambda_T=lambda_T;
            obj.phi_T=phi_T;
            obj.rho_bar=rho_bar;
        end
    end

    methods
        % Claculate methos
        function t = gen_timeline(obj)
            % Generate a timeline for simulation
            %
            % Returns:
            % t: （row vector)timeline for simulation.

            T = obj.days*24*60*60; % totlal times of simulation (s)
            tmax = T/obj.h; % time index

            t=linspace(0,(tmax-1)*obj.h,tmax);
        end
        function [lambda,phi]= cal_nadir_point(obj,theta_k)
            % Calculate the track of substellar point
            omega_e = 7.292e-5;
            t = gen_timeline(obj);
            % Convert parameters from deg to rad 
            M_rad=deg2rad(obj.M);
            theta_k_rad=deg2rad(theta_k);
            i_rad=deg2rad(obj.i);
            Omega_rad=deg2rad(obj.Omega);
            G_0_rad=deg2rad(obj.G_0);

            % Calculate M_K and Omega_k by rad.
            delta_M_temp=deg2rad(obj.b_M.*t);
            delta_Omega_temp=deg2rad(obj.b_Omega.*t);
            M_k=M_rad+theta_k_rad+delta_M_temp;
            Omega_k=Omega_rad+delta_Omega_temp;

            % Calculate lambda and phi by rad.
            lambda=Omega_k+atan(tan(M_k).*cos(i_rad))-(G_0_rad+omega_e.*t);
            phi=asin(sin(M_k).*sin(i_rad));
        end
        function obj = cal_tau(obj, theta_k)
            arguments
                obj 
                theta_k(1,1) double =0;
            end
            R_e=6371.393e3;
            [lambda,phi]= cal_nadir_point(obj,theta_k);
            lambda_T_rad=deg2rad(obj.lambda_T);
            phi_T_rad=deg2rad(obj.phi_T);
            rho_bar_rad=deg2rad(obj.rho_bar);
            % Calculate target Earth-Fixed Coordinate System.
            x_T=R_e*cos(lambda_T_rad)*cos(phi_T_rad);
            y_T=R_e*sin(lambda_T_rad)*cos(phi_T_rad);
            z_T=R_e*sin(phi_T_rad);

            % Calculate substellar point track in the earth-fixed coordinate system.
            x_k=R_e.*cos(lambda).*cos(phi);
            y_k=R_e.*sin(lambda).*cos(phi);
            z_k=R_e.*sin(phi);


            % Calculate the ground coverage time window.
            tempT=[x_T;y_T;z_T];
            tempK=[x_k',y_k',z_k'];
            rho=zeros(size(tempK,1),1);

            dotProducts=tempT' *tempK';
            norm_tempT=norm(tempT);
            norms_tempK=sqrt(sum(tempK.^2,2));

            rho=acos(dotProducts ./(norm_tempT .* norms_tempK'));
            obj.tau=(abs(rho)<rho_bar_rad);
        end
    end
end



